--- model_reader.py.orig	2026-03-16 13:33:35
+++ model_reader.py	2026-03-16 13:34:17
@@ -65,6 +65,7 @@
         couplings, plus add new dictionary coupling_dict from
         parameter name to value."""
 
+        all_params = {}
         param_card_text = None
         # Extract external parameters
         external_parameters = self['parameters'][('external',)]
@@ -201,7 +202,7 @@
                                 raise
                             
                     exec("locals()[\'%s\'] = %s" % (parameter_dict[block][pid].name,
-                                          value))
+                                          value), globals(), all_params)
                     parameter_dict[block][pid].value = float(value)
            
         else:
@@ -210,14 +211,14 @@
                 if scale and parameter_dict[block][id].name == 'aS':
                     runner = Alphas_Runner(value, nloop=3)
                     value = runner(scale)
-                exec("locals()[\'%s\'] = %s" % (param.name, param.value))
+                exec("locals()[\'%s\'] = %s" % (param.name, param.value), globals(), all_params)
 
-            
+
         # Define all functions used
         for func in self['functions']:
             exec("def %s(%s):\n   return %s" % (func.name,
                                                 ",".join(func.arguments),
-                                                func.expr))
+                                                func.expr), globals(), all_params)
 
         # Extract derived parameters
         derived_parameters = []
@@ -230,12 +231,12 @@
         # Now calculate derived parameters
         for param in derived_parameters:
             try:
-                exec("locals()[\'%s\'] = %s" % (param.name, param.expr))
+                exec("locals()[\'%s\'] = %s" % (param.name, param.expr), globals(), all_params)
             except Exception as error:
                 msg = 'Unable to evaluate %s = %s: raise error: %s' % (param.name,param.expr, error)
                 raise MadGraph5Error(msg)
-            param.value = complex(eval(param.name))
-            if not eval(param.name) and eval(param.name) != 0:
+            param.value = complex(all_params[param.name])
+            if not all_params[param.name] and all_params[param.name] != 0:
                 logger.warning("%s has no expression: %s" % (param.name,
                                                              param.expr))
 
@@ -244,18 +245,18 @@
         for particle in self.get('particles'):
             if particle.is_fermion() and particle.get('self_antipart') and \
                    particle.get('width').lower() != 'zero' and \
-                   eval(particle.get('mass')).real < 0:
+                   eval(particle.get('mass'), globals(), all_params).real < 0:
                 exec("locals()[\'%(width)s\'] = -abs(%(width)s)" % \
-                     {'width': particle.get('width')})
+                     {'width': particle.get('width')}, globals(), all_params)
 
         # Extract couplings
         couplings = sum(list(self['couplings'].values()), [])
         # Now calculate all couplings
         for coup in couplings:
             #print "I execute %s = %s"%(coup.name, coup.expr)
-            exec("locals()[\'%s\'] = %s" % (coup.name, coup.expr))
-            coup.value = complex(eval(coup.name))
-            if not eval(coup.name) and eval(coup.name) != 0:
+            exec("locals()[\'%s\'] = %s" % (coup.name, coup.expr), globals(), all_params)
+            coup.value = complex(all_params[coup.name])
+            if not all_params[coup.name] and all_params[coup.name] != 0:
                 logger.warning("%s has no expression: %s" % (coup.name,
                                                              coup.expr))
 
@@ -270,7 +271,7 @@
         self.set('coupling_dict', dict([(coup.name, coup.value) \
                                         for coup in couplings]))
         
-        return locals()
+        return all_params
     
     def get_mass(self, pdg_code):
         """easy way to have access to a mass value"""
